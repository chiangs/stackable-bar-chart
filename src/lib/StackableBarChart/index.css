/* Temp styles */

/* #region Utils */

/* #endregion Utils */

/* #region Container */
.stackable-container {
  --font-size: clamp(1rem, 5vw, 60px);
  --radius: 15px;
  --radius-tooltip: 4px;
  --radius-bar: var(--radius) 0 var(--radius) 0;
  --color-fallback: cadetblue;
}

.stackable-container * {
  box-sizing: border-box;
}

.chart-title {
  margin-top: 20px;
  width: 100%;
  text-align: center;
}

.data-label {
  color: var(--color-fallback);
  white-space: nowrap;
  text-align: right;
}

.data-label h3,
.bar-value {
  padding: 0;
  margin: 0;
  font-size: var(--font-size);
}

/* #region stacked */
.stackable-container.stacked {
  display: flex;
  justify-content: space-between;
  flex-direction: column;
}

.stackable-container.stacked .chart-container {
  background: transparent;
  min-width: 350px;
  height: 100%;
  margin: 0;
  display: grid;
  grid-gap: 20px;
  grid-template-rows: 1fr;
  grid-template-columns: 1fr 2fr;
  position: relative;
}

.stacked .data-label,
.stacked .data-bar {
  display: flex;
  align-items: stretch;
}

.stacked .data-label {
  align-items: center;
  /* justify-content: flex-end; */
}

.stacked.title-bottom {
  flex-direction: column;
}

.stacked.title-top,
.stacked.title-default {
  flex-direction: column-reverse;
}

.stacked.title-left {
  flex-direction: row-reverse;
}

.stacked.title-right {
  flex-direction: row;
}

.title-default .chart-title {
  text-align: left;
  margin: 0;
}

.title-default .chart-title h2 {
  margin-top: 0;
}

.stacked.title-left .chart-title,
.stacked.title-right .chart-title {
  width: initial;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: stretch;
}

.stackable-bar.stacked,
.stacked .bar-value {
  border-radius: var(--radius-bar);
}
/* #endregion stacked */

/* #region linear */
.stackable-container.linear {
  display: flex;
  flex-direction: column;
}

.stackable-container.linear .chart-bars {
  width: 100%;
  display: flex;
  justify-content: flex-start;
}

.linear .stackable-bar {
  height: 125px;
  position: relative;
}
/* #endregion linear */
/* #endregion Container */

/* #region Bar */
.stackable-bar {
  cursor: default;
  width: 100%;
  display: flex;
  align-items: center;
  overflow: hidden;
}

.stackable-bar.tooltip {
  transform: scale(1);
  transition: all 0.1s ease-in;
}

.stackable-bar.linear.tooltip:hover {
  transform: scale(1.1);
  transition: all 0.15s ease-out;
  z-index: 100;
}

/* #region tooltip */
.tooltip-container {
  z-index: 100;
  position: absolute;
  margin: 0;
  padding: 5px;
  background-color: white;
  border-radius: var(--radius-tooltip);
  box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.65), 0px 4px 8px rgba(0, 0, 0, 0.75);
}

/* #endregion tooltip */
.bar-value {
  width: 0;
  height: 100%;
  margin: 0;
  padding: 5px;
  font-size: var(--font-size);
  font-weight: bold;
  background: var(--color-fallback);
}

.data-label,
.bar-value-text {
  user-select: all;
}

.stacked-content {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: stretch;
}

.stacked-content * {
  margin: 0;
}

.stacked-content .bar-value-text {
  font-size: clamp(1.5rem, 3vw, 60px);
}

.bar-label-text {
  font-size: clamp(1.3rem, 1.5vw, 60px);
  white-space: nowrap;
}
/* #endregion Bar */
