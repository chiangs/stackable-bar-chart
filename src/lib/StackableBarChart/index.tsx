import React, { useState } from "react";
import "./index.css";
import type {
  BarData,
  BarProps,
  ChartProps,
  Rounding,
  SortProperty,
} from "../__types";
import Bar from "../Bar";
import Label from "../Label";
import Tooltip from "../Tooltip";
import ChartContainer from "../ChartContainer";
type Props = ChartProps;

const NAME_COMPONENT = "stackable-container";

/**
 * Sorts collection
 * @param data
 * @param sortedBy
 * @returns sorted collection from largest or smallest
 */
const sortDataSmallBig = (
  data: BarProps[],
  sortedBy: SortProperty = "none"
): BarProps[] =>
  data.sort((a, b) =>
    sortedBy === "smallest" ? a.value - b.value : b.value - a.value
  );

/**
 * Rounding helper
 * @param value
 * @param method
 * @returns value rounded
 */
const roundPortion = (value: number, method: Rounding): number => {
  if (method === "up") {
    return Math.ceil(value);
  }
  if (method === "down") {
    return Math.floor(value);
  }
  return Math.round(value);
};

/**
 * Calculates the actual portion and render portion
 * @param data
 * @returns BarProps[] with portion calculations
 */
const calcPortionsForData = (
  data: BarProps[],
  method: Rounding,
  sortedBy: SortProperty = "none"
): BarProps[] => {
  // CB - actual percentage
  const calcPortion = (value: number, total: number, method: Rounding) => {
    const portion = (value / total) * 100;
    return roundPortion(portion, method);
  };
  // CB - percentage for rendering visual size
  const calcRenderPortion = (value: number, largest: number) =>
    (value / largest) * 100;
  // CB - Identifies largest value in collection
  const findLargest = (collection: BarProps[]) => {
    let temp = 0;
    let largest = collection[0];
    collection.forEach((d) => {
      if (temp < d.value) {
        temp = d.value;
        largest = d;
      }
    });
    return largest;
  };
  // largest used as base of 100% of container
  const largest =
    sortedBy === "none"
      ? findLargest(data)
      : sortedBy === "smallest"
      ? data[data.length - 1]
      : data[0];
  const sum = data.reduce((t, o) => {
    return t + o.value;
  }, 0);
  const updated = data.map((d) => ({
    ...d,
    portion: calcPortion(d.value, sum, method),
    renderPortion: calcRenderPortion(d.value, largest.value),
  }));
  return updated;
};

// TODO: Optimize CSS file
// TODO: Integration tests
const StackableBarChart: React.FC<Props> = ({
  data = [],
  sortLinear = "none",
  mode = "linear",
  rounding = "nearest",
  colorBackground = "#fff",
  showPercentage = false,
  showTooltip = true,
  titlePosition = "default",
  children,
  clickHandler = () => null,
}) => {
  const [tooltipContent, setTooltipContent] = useState<BarData | null>();

  // Events
  const onHoverBar = (data: any | null) => setTooltipContent(data);
  const onClickBar = (data: Partial<BarData>) => clickHandler(data);
  const onBlurFigure = () => setTooltipContent(null);

  // Get background of app for knockout bar value text
  const background =
    colorBackground === "transparent" ? "#fff" : colorBackground;

  // Process data collection
  const sortedData =
    sortLinear === "none" ? data : sortDataSmallBig(data, sortLinear);
  const sortedDataWithPortion = calcPortionsForData(
    sortedData,
    rounding,
    sortLinear
  );

  // Bars
  let bars;
  if (mode === "stacked") {
    bars = sortedDataWithPortion.map((d, i) => (
      <React.Fragment key={`${d.label}${i}`}>
        <div className="data-label" style={{ color: d.color }}>
          <Label>{d.label}</Label>
        </div>
        <div className="data-bar">
          <Bar
            {...d}
            background={background}
            mode={mode}
            showTooltip={showTooltip}
            revealTooltipHandler={onHoverBar}
            barClickHandler={onClickBar}
          />
        </div>
      </React.Fragment>
    ));
  } else {
    bars = sortedDataWithPortion.map((d, i) => (
      <React.Fragment key={`${d.label}${i}`}>
        <Bar
          {...d}
          background={background}
          mode={mode}
          showPercentage={showPercentage}
          showTooltip={showTooltip}
          revealTooltipHandler={onHoverBar}
          barClickHandler={onClickBar}
        />
      </React.Fragment>
    ));
  }

  // Title
  let title = null;
  if (titlePosition !== "none") {
    title = children;
  }

  // Tooltip
  const tooltip = showTooltip && tooltipContent && (
    <Tooltip {...tooltipContent} />
  );

  // Title
  const chartTitle = <figcaption className="chart-title">{title}</figcaption>;

  // ChartContents
  const contents =
    mode === "stacked" ? (
      <>
        {bars}
        {tooltip}
      </>
    ) : (
      <>
        {" "}
        <div className="chart-bars">{bars}</div>
        {tooltip}
      </>
    );

  //  Chart
  const chart = <ChartContainer>{contents}</ChartContainer>;

  return (
    <figure
      className={`${NAME_COMPONENT} ${mode} title-${titlePosition}`}
      onBlur={onBlurFigure}
      data-testid={NAME_COMPONENT}
    >
      {chart}
      {chartTitle}
    </figure>
  );
};

export default StackableBarChart;
