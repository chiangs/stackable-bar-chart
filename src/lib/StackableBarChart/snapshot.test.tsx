import * as React from "react";
import renderer from "react-test-renderer";
import { describe, expect, test } from "vitest";
import StackableBarChart from ".";

describe("StackableBarChart", () => {
  test("StackableBarChart component renders correctly", () => {
    const component = renderer.create(<StackableBarChart />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
