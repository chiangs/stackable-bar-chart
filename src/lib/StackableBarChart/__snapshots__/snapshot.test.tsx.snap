// Vitest Snapshot v1

exports[`StackableBarChart > StackableBarChart component renders correctly 1`] = `
<figure
  className="stackable-container stacked title-bottom"
  data-testid="stackable-container"
>
  <div
    className="chart-container"
  />
  <div
    className="chart-title"
  />
</figure>
`;
